File: Algorithm/074. Search a 2D Matrix.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-14 14:38:29
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-14 14:38:37
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        m = len(matrix)
        if m>0:
            n = len(matrix[0])
            if n == 0:
                return False
        else:
            return False
        
        left, right = 0, m-1
        while left < right:
            mid = (left + right) / 2
            if matrix[mid][0] == target:
                return True
            elif matrix[mid][0] > target:
                right = mid - 1
            elif matrix[mid][0] < target <= matrix[mid][n-1]:
                left = mid
                break
            else:
                left = mid+1
        r = left 
        left, right = 0, n-1
        while left < right:
            mid = (left+right)/2
            if matrix[r][mid] == target:
                return True
            elif matrix[r][mid] > target:
                right = mid - 1
            elif matrix[r][mid] < target:
                left = mid + 1
        if matrix[r][left] == target:
            return True
        else:
            return False

File: Algorithm/Java/074. Search a 2D Matrix.java
/**
* Author: WuLC
* Date:   2017-01-15 17:07:14
* Last modified by:   WuLC
* Last Modified time: 2017-01-15 17:08:47
* Email: liangchaowu5@gmail.com
*/


// compare target with the number on top right corner, and decide whether to delete the row or the column
public class Solution 
{
    public boolean searchMatrix(int[][] matrix, int target) 
    {
        int m = matrix.length;
        if (m==0) return false;
        int n = matrix[0].length;
        int row = 0,  col = n-1;
        while (row < m && col >= 0)
        {
            if (matrix[row][col] == target) return true;
            else if(matrix[row][col] < target) row++;
            else col--;
        }
        return false;
    }
}